src/ripple/app/misc/impl/ValidatorList.cpp

```diff
@@ -395,9 +424,94 @@ ValidatorList::removePublisherList (PublicKey const& publisherKey)
             --iVal->second;
     }
 
+    iList->second.list.clear();
+    iList->second.available = false;
+
     return true;
 }
 
+boost::optional<TimeKeeper::time_point>
+ValidatorList::expires() const
+{
+    boost::shared_lock<boost::shared_mutex> read_lock{mutex_};
+    boost::optional<TimeKeeper::time_point> res{boost::none};
+    for (auto const& p : publisherLists_)
+    {
+        if (p.second.available &&
+            TimeKeeper::time_point{} < p.second.expiration)
```

Comment:
I'm wondering if this should return `boost::none` if a list has a `TimeKeeper::time_point{}` expiration (meaning it hasn't been fetched yet).
As is, we have this subset of possible `validator_list_expires` values:
-unfetched list: `unknown`
-unfetched list & one or more fetched lists: the timestamp of the soonest expiration
-unfetched list & static validator keys: `never`

I'm leaning toward always making `validator_list_expires` to be `unknown` if any list has yet to be successfully fetched.

Reply:
Even though the quorum becomes infinite if a list is unavailable, don't you still process/forward validations for whatever trusted nodes are available? If that is the case, I think providing the expiration of the trusted keys you are using is still useful.

But I don't hold that opinion particularly strongly, so if you prefer the above, I will switch.

Reply:
We'd still have the `expiration` in each `publisher_lists` entry for the `validators` response to tell us when we might stop trusting and forwarding certain validators.
I've been thinking of `validator_list_expires` as indicating the earliest point at which you might have an incomplete trusted validator list and be unable to reach quorum, but we would need to return something like `unknown` whenever if we already have an incomplete list.

Reply:
👍  I'll make the change then.

Reply:
On a second take, I'm not sure why I was using the `available` field in this function. That looks to be used to indicate whether a list has been loaded and is unexpired, but if it expires, we don't want to stop reporting its expiry.

Is the better change to just
* Return `boost::none` if any `expiration` is `TimeKeeper::time_point{}`
* Otherwise, return the earliest `expiration`.

Reply:
Good catch. Changing `available` back to false when the list expires was just added in https://github.com/ripple/rippled/pull/2240
Your suggested better change looks correct.

Reply:
It might be worth adding a test to `testExpires` and/or `testDynamicUNL` where an applied list expires
https://github.com/ripple/rippled/blob/cafe18c59279e7de447f25a0e00d0562d6441c7a/src/test/app/ValidatorList_test.cpp#L745-L746

Reply:
👍  Already working on one in `testExpires`.